src/hotspot/cpu/aarch64/pauth_aarch64.hpp

```diff
@@ -25,11 +25,121 @@
+// should ensure new values are correctly signed. However, we cannot make any assumptions about
+// how (or if) native code uses PAC - here we should limit access to viewing via stripping.
+//
+
+// Use only the PAC instructions in the NOP space. This ensures the binaries work on systems
+// without PAC. Write these instructions using their alternate "hint" instructions to ensure older
+// compilers can still be used. For Apple, instead use the recommended pauth interface.
+#define XPACLRI   "hint #0x7;"
+#define PACIA1716 "hint #0x8;"
+#define AUTIA1716 "hint #0xc;"
+
+
+// Strip an address. Use with caution - only if there is no guaranteed way of authenticating the
+// value.
+//
+inline address pauth_strip_pointer(address ptr) {
+#ifdef __APPLE__
+  return ptrauth_strip(ptr, ptrauth_key_asib);
+#else
+  register address result __asm__("x30") = ptr;
+  asm (XPACLRI : "+r"(result));
+  return result;
+#endif
+}
+
+// Confirm the given pointer has not been signed - ie none of the high bits are set.
+//
+// Note this can give false positives. The PAC signing can generate a signature with all signing
+// bits as zeros, causing this function to return true. Therefore this should only be used for
+// assert style checking. In addition, this function should never be used with a "not" to confirm
+// a pointer is signed, as it will fail the above case. The only safe way to do this is to instead
+// authenticate the pointer.
+//
 inline bool pauth_ptr_is_raw(address ptr) {
-  // Confirm none of the high bits are set in the pointer.
   return ptr == pauth_strip_pointer(ptr);
 }
 
+// Sign a return value, using the given modifier.
+//
+inline address pauth_sign_return_address(address ret_addr, address modifier) {
+  if (UseROPProtection) {
+    // A pointer cannot be double signed.
+    assert(pauth_ptr_is_raw(ret_addr), "Return address is already signed");
+#ifdef __APPLE__
+    ret_addr = ptrauth_sign_unauthenticated(ret_addr, ptrauth_key_asib, modifier);
+#else
+    register address r17 __asm("r17") = ret_addr;
+    register address r16 __asm("r16") = modifier;
+    asm volatile (PACIA1716 : "+r"(r17) : "r"(r16));
+    ret_addr = r17;
+#endif
+  }
+  return ret_addr;
+}
+
+// Authenticate a return value, using the given modifier.
+//
+inline address pauth_authenticate_return_address(address ret_addr, address modifier) {
+  if (UseROPProtection) {
+#ifdef __APPLE__
+    ret_addr = ptrauth_auth_data(ret_addr, ptrauth_key_asib, modifier);
+#else
+    register address r17 __asm("r17") = ret_addr;
+    register address r16 __asm("r16") = modifier;
+    asm volatile (AUTIA1716 : "+r"(r17) : "r"(r16));
+    ret_addr = r17;
+#endif
+    // Ensure that the pointer authenticated.
+    assert(pauth_ptr_is_raw(ret_addr), "Return address did not authenticate");
+  }
+  return ret_addr;
+}
+
+// Authenticate or strip a return value. Use for efficiency and only when the safety of the data
+// isn't an issue - for example when viewing the stack.
+//
```

Comment:
So, whether this function authenticates or strips the address depends only on debugging? The vague name makes the callers hard to read.

Reply:
>whether this function authenticates or strips the address depends only on debugging?

Yes. We only need to strip the value, because we're not jumping to the lr value, only viewing it.

The interface is different to a strip (as we need to pass in the modifier). 

How about something like pauth_authenticate_fast() ? or pauth_authenticate_unsafe() ?

Alternatively, this function is only called by the functions in Frame, so the frequency of use is probably low enough (compared to the sign/auth every function) that it's not going to cause any performance issues. So, could just replace with calls to pauth_authenticate. I think that might be the best option.

Reply:
A simple rule here: function names go with what the release version does. So I'd go with the actual purpose, which is `pauth_strip_addr_for_debuginfo()`. That's right, isn't it? You only want this thing for stack traces, logs, etc.

Reply:
This function is used by the frame code. So, that means it is used for all stack walks which are far from being simply cosmetic/ornamental. The runtime will rely on this for various different types of thread housekeeping.

The difference here is that in product mode this simply strips auth bits whereas in debug mode it actually authenticates as it strips to give extra verification. So, your suggested name is quite misleading. Likewise Alan's suggested names is misleading because the primary product operation is to strip not authenticate.

How about pauth_strip_verifiable? and a comment saying that it differs from pauth_strip by actually authenticating when debug is enabled.